sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java

```diff
@@ -67,6 +67,13 @@ public LuceneReferenceResolver(FieldTypeLookup fieldTypeLookup, IndexSettings in
     public LuceneCollectorExpression<?> getImplementation(Reference refInfo) {
         assert refInfo.granularity() == RowGranularity.DOC : "lucene collector expressions are required to be on DOC granularity";
 
+        if (refInfo.valueType().id() == ObjectType.ID && refInfo.granularity() == RowGranularity.DOC) {
```

Comment:
There is an assertion for granularity == DOC just 2 lines above?

What about https://github.com/crate/crate/blob/b00f1e4d95e18c6a827084c86eacddec3aca044c/sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java#L133

why is that not working?

Reply:
>There is an assertion for granularity == DOC just 2 lines above?

Yes, and?

The regular object lookup doesn't work because the reference has not been rewritten to a doc source lookup. This is done here:

https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/planner/operators/FetchOrEval.java#L336

>why is that not working?

Because the `Reference` points to `doc.table->obj` which is not an actual field in the field lookup. It needs to be `doc.table->_doc['obj']`.

Reply:
Sorry guess I should've been more specific.

I wanted to know what's up with https://github.com/crate/crate/blob/b00f1e4d95e18c6a827084c86eacddec3aca044c/sql/src/main/java/io/crate/operation/reference/doc/lucene/LuceneReferenceResolver.java#L133

because if you add the ObjectType.ID check like in this PR, the code below becomes effectively dead code as it will never be used. Which means the `ObjectColumnReference` is probably dead as well. 

Given that we always have to do source lookups for objects (if it's `obj` or `_doc['obj']` doesn't matter) , we could probably use the same expression implementation all the time.

Reply:
@mfussenegger I've replaced `ObjectColumnReference` and `GeoShapeColumnReference` with the doc source lookup. Seems to work fine. I wonder, dos this come at a performance cost because the lookup in the source is more expensive?